numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1843,19 +1879,38 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj)
         else {
             unsigned char mask;
 
-            for (index = 0; index < in_n; index++) {
-                for (mask = 128; mask; mask >>= 1) {
+            if (order == 'b') {
+                for (index = 0; index < in_n; index++) {
+                    for (mask = 128; mask; mask >>= 1) {
+                        *outptr = ((mask & (*inptr)) != 0);
+                        outptr += out_stride;
+                    }
+                    inptr += in_stride;
+                }
+                /* Clean up the tail portion */
+                mask = 128;
+                for (i = 0; i < in_tail; i++) {
                     *outptr = ((mask & (*inptr)) != 0);
                     outptr += out_stride;
+                    mask >>= 1;
                 }
-                inptr += in_stride;
             }
-            /* Clean up the tail portion */
-            mask = 128;
-            for (i = 0; i < in_tail; i++) {
-                *outptr = ((mask & (*inptr)) != 0);
-                outptr += out_stride;
-                mask >>= 1;
+            else {
+                for (index = 0; index < in_n; index++) {
+                    /* stops when char(128)<<=1 overflows to 0 */
+                    for (mask = 1; mask; mask <<= 1) {
```

Comment:
There's no guarantee that `char(256)` overflows, `CHAR_BIT` might not be 8.

Why not just use a simple loop here?

```C
for (int i = 0; i < 8; i++) {
    *outptr = (bool)(*inptr & (1 << i));
    outptr += out_stride;
}
```

Reply:
adopted